Add `siteID` and `keyword` attributes to `Customer` and `CustomerSearchResult` entities

iamgabrielma opened the pull request:
Part of #7741 

Apologies for the ~4.5k diff, most of it comes from regenerating the data entities, and the mapping model migration from 74 to 75, when adding `siteID` to the entities that previously didn't have them.

### Description
When we created the Storage layer and data model for `Customer` and `CustomerSearchResult` on https://github.com/woocommerce/woocommerce-ios/pull/7841, we (I) didn't realize that we share the same storage across all sites a merchant may have (we do not hold 1 different database per site, but 1 per all sites). This may cause unintended consecuences when a merchant has more than 1 store in their account, as we wouldn't have a way to filter `customerID` per site basis when we store fetched customers and results.

### Changes

This PR updates the `Customer` and `CustomerSearchResult` by adding a siteID to both entities, as well as substituting `customerID` attribute for the `keyword` attribute in the `CustomerSearchResult` entity.

- Updated Data Model from 74 to 75.
- Add non-optional siteID property to `Customer` and `CustomerSearchResults`
- Add non-optional `keyword` property to `CustomerSearchResult`
- Remove `customerID` property from `CustomerSearchResult`
- Used mapping model: `WooCommerceModelV74toV75.xcmappingmodel` to remove `Customer` and `CustomerSearchResults` entities in model version 74
- Added a `74_to_75` test to MigrationTests to confirm that the properties are correct after the data model migration.
- Updated `MIGRATIONS.md`

### Testing instructions
- Confirm that Unit tests pass
- Confirm that the app doesn't crash

iamgabrielma committed:
Update data model to 75

Adds siteID attributes to Customer & CustomerSearchResults. Currently the migration test 74_75 is failing when attempting to migrate()

iamgabrielma commented on Storage/StorageTests/CoreData/MigrationTests.swift:
```diff
@@ -1498,14 +1517,29 @@ private extension MigrationTests {
     @discardableResult
-    func insertCustomerSearchResult(to context: NSManagedObjectContext) -> NSManagedObject {
-        context.insert(entityName: "CustomerSearchResult", properties: [
+    func insertCustomerSearchResult(to context: NSManagedObjectContext, forModel modelVersion: Int) -> NSManagedObject {
+        let customerSearchResult = context.insert(entityName: "CustomerSearchResult", properties: [
             "customerID": 1
         ])
+
+        // Required since model 75
+        if modelVersion >= 75 {
+            customerSearchResult.setValue(1, forKey: "siteID")
+            customerSearchResult.setValue("", forKey: "keyword")
+        }
+
+        return customerSearchResult
```
Here I forgot to remove the `customerID` in the new model (75), as doesn't exist anymore, but changing the logic didn't resolve the problem of the `migrate()` method failing. For reference I changed it to:

```
    /// Inserts a `CustomerSearchResult` entity, providing default values for the required properties.
    @discardableResult
    func insertCustomerSearchResult(to context: NSManagedObjectContext, forModel modelVersion: Int) -> NSManagedObject {
        if modelVersion >= 75 {
            let customerSearchResult = context.insert(entityName: "CustomerSearchResult", properties: [
                "siteID": 1,
                "keyword": "hi"
            ])
            return customerSearchResult
        } else {
            let customerSearchResult = context.insert(entityName: "CustomerSearchResult", properties: [
                "customerID": 1
            ])
            return customerSearchResult
        }
    }
```

iamgabrielma committed:
Make keyword attribute optional. fix migration tests

iamgabrielma committed:
Update migrations.md

iamgabrielma commented on Storage/StorageTests/CoreData/MigrationTests.swift:
```diff
@@ -1498,14 +1529,30 @@ private extension MigrationTests {
-        context.insert(entityName: "CustomerSearchResult", properties: [
-            "customerID": 1
-        ])
+    func insertCustomerSearchResult(to context: NSManagedObjectContext, forModel modelVersion: Int) -> NSManagedObject {
+        if modelVersion < 75 {
+            let customerSearchResult = context.insert(entityName: "CustomerSearchResult", properties: [
+                "customerID": 1])
+            return customerSearchResult
+        } else {
+            // Required since model 75
+            let customerSearchResult = context.insert(entityName: "CustomerSearchResult", properties: [
+                "siteID": 1,
+                "keyword": ""
+            ])
+            return customerSearchResult
+        }
```
Interestingly, here I had to use this logic rather than the one I used on `insertCustomer(to:forModel:)`, as we're removing the `customerID` attribute entirely, migration tests were failing due to incompatibility between models. Happy to refactor if there's a better way to test this.

joshheald replied:
Makes sense. You might just want to do the old version in-line in the test which uses it, since it'll never be relevant again, but up to you on that one...

iamgabrielma replied:
Much more cleaner indeed. Changed here 312e9fd466f620394766d2c56c97d52b679d4480 and removed the now unused helper function here 761736be445ad848f631b91366dfe06a1e0657c8

joshheald commented on Storage/Storage/Model/Customer+CoreDataProperties.swift:
```diff
@@ -8,34 +8,34 @@ extension Customer {
+    @NSManaged public var billingCompany: String?
     @NSManaged public var billingCountry: String?
-    @NSManaged public var billingPhone: String?
     @NSManaged public var billingEmail: String?
-    @NSManaged public var shippingFirstName: String?
-    @NSManaged public var shippingLastName: String?
-    @NSManaged public var shippingCompany: String?
+    @NSManaged public var billingFirstName: String?
+    @NSManaged public var billingLastName: String?
+    @NSManaged public var billingPhone: String?
+    @NSManaged public var billingPostcode: String?
+    @NSManaged public var billingState: String?
+    @NSManaged public var customerID: Int64
+    @NSManaged public var email: String?
+    @NSManaged public var firstName: String?
+    @NSManaged public var lastName: String?
     @NSManaged public var shippingAddress1: String?
     @NSManaged public var shippingAddress2: String?
     @NSManaged public var shippingCity: String?
-    @NSManaged public var shippingState: String?
-    @NSManaged public var shippingPostcode: String?
+    @NSManaged public var shippingCompany: String?
     @NSManaged public var shippingCountry: String?
-    @NSManaged public var shippingPhone: String?
     @NSManaged public var shippingEmail: String?
+    @NSManaged public var shippingFirstName: String?
+    @NSManaged public var shippingLastName: String?
+    @NSManaged public var shippingPhone: String?
+    @NSManaged public var shippingPostcode: String?
+    @NSManaged public var shippingState: String?
+    @NSManaged public var siteID: Int64
```
Yay inconsistent code generation... this isn't a problem, but this is what I meant when I mentioned that I'll often just run the generation to check it works, then revert it and add/update properties on the generated files so that the change is easier to see. No need to do anything, but it's a good practical example.

joshheald commented on Storage/Storage/Model/CustomerSearchResult+CoreDataProperties.swift:
```diff
@@ -8,9 +8,9 @@ extension CustomerSearchResult {
 
-    @NSManaged public var customerID: Int64
+    @NSManaged public var siteID: Int64
+    @NSManaged public var keyword: String?
```
> Add optional keyword property to CustomerSearchResult

Why optional? I don't think it makes sense to have a search result with a nil keyword...

iamgabrielma replied:
🙇  Made the changes discussed in Slack here: 312e9fd466f620394766d2c56c97d52b679d4480 by changing `keyword` to be non-optional and providing a default value so we're sure that the entity conforms to the new schema when migrating

joshheald reviewed (approved):
Looks good... a question about the optional `keyword`, and some general observations, but feel free to merge with or without the optional change.

iamgabrielma committed:
Make keyword non-optional. Simplify test helper

Since we’re making the “keyword” attribute non-optional, we’re also providing a default value to the CustomerSearchResult entity so we can migrate between data models without issues, as any existing CustomerSearchResult would need to have a “keyword” set to it when we’re migrating data sets. With this we also altered the 74_to_75 tests to check migrated entities between data models, rather than new entities in the new data model.

iamgabrielma committed:
Remove unneeded helper function

iamgabrielma commented on Storage/Storage/Model/MIGRATIONS.md:
```diff
@@ -2,6 +2,13 @@
 
 This file documents changes in the WCiOS Storage data model. Please explain any changes to the data model as well as any custom migrations.
 
+## Model 75 (Release 10.8.0.0)
```
Needs to be updated

itsmeichigo reviewed (commented):
❓ Looking at the changes here, I think we won't have the correct `siteID` for `Customer` and `CustomerSearchResult` after the migration between 74 and 75 - since the attributes didn't exist in version 74 (the default values will be 0). Do you think we should do heavyweight migration to handle the missing attributes instead?

iamgabrielma committed:
Add WooCommerceModelV74toV75 mapping model

iamgabrielma committed:
Fix line length warning

iamgabrielma commented:
> ❓ Looking at the changes here, I think we won't have the correct `siteID` for `Customer` and `CustomerSearchResult` after the migration between 74 and 75 - since the attributes didn't exist in version 74 (the default values will be 0). Do you think we should do heavyweight migration to handle the missing attributes instead?

Thanks for your help on this! In https://github.com/woocommerce/woocommerce-ios/pull/7860/commits/db370f8898e8d54d4e7876a5efdf45b78e6163b2 a new `WooCommerceModelV74toV75` has been added as well as new tests to `test_migrating_from_74_to_75_adds_siteID_and_keyword_attributes_to_Customer_and_CustomerSearchResult` so we can confirm the changes to both entities and attributes, pre and post-migration.

iamgabrielma commented:
Quick update: There's a step missing and tests were passing because of an error here: https://github.com/woocommerce/woocommerce-ios/pull/7860/files#diff-03c54fe20fd9822d0011069f12658bf05260422a448742db92b81c339a6882deR1418-R1419

Currently, in the MigrationTests we're checking if the entities deletion happens in the `sourceContext`, which is wrong in this step as we should be checking the `targetContext` instead:

```
- XCTAssertEqual(try sourceContext.count(entityName: "Customer"), 0)
- XCTAssertEqual(try sourceContext.count(entityName: "CustomerSearchResult"), 0)
+ XCTAssertEqual(try targetContext.count(entityName: "Customer"), 0)
+ XCTAssertEqual(try targetContext.count(entityName: "CustomerSearchResult"), 0)
```

The missing step is to use `DeleteEntityObjectsMigrationPolicy` in the Mapping Model to be sure that this entity deletion happens between models. This, and the tests, have been fixed on https://github.com/woocommerce/woocommerce-ios/pull/7860/commits/07ae1ca83e0cf23408f88608f25dba040f0cf10e

iamgabrielma committed:
Add NSEntityMigrationPolicy to mapping migration

itsmeichigo reviewed (approved):
Great work 🎉
